Fix blank user name

jerinkc opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://developers.forem.com/contributing-guide/forem#create-a-pull-request
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
- Presence validation for name and corresponding fixes in the UI

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #19617 (does not close due to remaining invalid data)

## QA Instructions, Screenshots, Recordings
1. Update the "Name" field of profile settings (settings -> profile)
2. Submit the form
Result: Displays error message "Name can't be blank" message.

Recording: https://www.loom.com/share/90c1e18bdb7a4621a03888d401d1317e?sid=67e95f38-8d56-4540-ac49-54d73528d02d

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_
- [ ] Semantic HTML implemented?
- [ ] Keyboard operability supported?
- [ ] Checked with [axe DevTools](https://www.deque.com/axe/) and addressed `Critical` and `Serious` issues?
- [ ] Color contrast tested?

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/forem/forem?pullRequest=19723) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/forem/forem?pullRequest=19723) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/forem/forem?pullRequest=19723) it.</sub>

rt4914 commented:
Thanks for the PR @jerinkc . @maestromac and @lightalloy will review this PR soon. Thanks.

PhilipHow commented:
@jerinkc, thanks so much for this PR - it's really appreciated.

> Redirecting users with blank name to profile settings page.

This is not something I was expecting as part of this change. Firstly my apologies for not being clearer in the issue to detail what we were expecting here - but in reality, it was only the validation part of the PR.

Could you remove this from your solution?

lightalloy reviewed (commented):
Hey, @jerinkc , thanks for the pr!
As we have decided that forcing users to update their `name` via redirect shouldn't be a part of this change, the changes related to displaying error messages are also not relevant.

The validation you have added is necessary, but it will leave some users in invalid state, I will discuss potential solutions to this problem with the team.

jerinkc commented:
Hey @lightalloy, Thanks for your feedback. Based on my understanding, you would like me to remove the error messages that appear on other tabs, but not the ones that show up on the form. If that is correct, I have already implemented this change and updated the code accordingly.

jerinkc commented:
Hey @lightalloy,

I apologize for the inadvertent changes related to the translation file. I appreciate your feedback and have reverted them in the latest commits.

Regarding the changes in `app/views/users/_errors.html.erb`, I have decided to keep them as they fix the current issue of displaying redundant and incorrect error messages, such as '1 error 1 error prohibited your profile from being saved::' or '2 errors 2 errors prohibited your profile from being saved::'.

I have also implemented the data update scripts to use username as the name, as you suggested. Thank you for this brilliant solution and documentation.

lightalloy commented on lib/data_update_scripts/20230718102644_backfill_user_blank_name.rb:
```diff
@@ -0,0 +1,10 @@
+  class BackfillUserBlankName
+    def run
+      User.where(" TRIM(name)='' ").each do |user|
+        titleized_username = user.username.titleize
```
I think we can go without titleizing, we allow names starting with a lowercase letter.

jerinkc replied:
No titleize, got it

lightalloy reviewed (commented):
Thanks for implementing the changes :sparkles: 
Looks good, I left one small comment about titleizing

lightalloy commented on app/views/users/_errors.html.erb:
```diff
@@ -1,7 +1,7 @@
   <div class="crayons-card__header">
     <h3 class="crayons-card__headline">
-      <%= pluralize(errors.size, "error") %> <%= title_suffix %>:
+      <%= title_suffix %>
```
Thanks for fixing the bug along the way :raised_hands:

maestromac reviewed (approved):
Thank you for this PR. LGTM!
